src/data/setoid/partition.lean

```diff
@@ -237,6 +244,89 @@ _ (subtype (@is_partition α)) _ (partial_order.to_preorder _) $ partition.order
 
 end partition
 
+section finpartition
+
+open_locale classical
+
+-- Variant of finset.mem_sup that does not assume that f is finset valued
+lemma finset.mem_sup' {α β : Type*}  {s : finset α} {f : α → set β}
+  {x : β} : x ∈ s.sup f ↔ ∃ v ∈ s, x ∈ f v :=
+begin
+  split,
+  { apply s.induction_on,
+    { intro hx,
+      exfalso,
+      simpa only using hx, },
+    intros a t hat hrec hx,
+    simp only [finset.sup_insert, set.sup_eq_union, set.mem_union_eq] at hx,
+    cases hx with hx hx,
+      exact ⟨a, finset.mem_insert_self a t, hx ⟩,
+      { obtain ⟨v, H, hH⟩ := hrec hx ,
+        exact ⟨v, finset.mem_insert_of_mem H, hH⟩, }},
+  { rintros ⟨v, H, hx⟩,
+    suffices : f v ≤ s.sup f,
+      exact this hx,
+    exact finset.le_sup H, }
+end
```

Comment:
I think this might be a better lemma:
```suggestion
lemma finset.sup_eq_bUnion {α β : Type*} {s : finset α} {f : α → set β} :
  s.sup f = ⋃ x ∈ s, f x :=
begin
  induction s using finset.cons_induction_on with a s ha ih,
  { simp },
  { simp [ih] },
end
```

Reply:
This one is good too:
```lean
lemma finset.sup_id_eq_sUnion {α β : Type*} {s : finset (set α)} :
  s.sup id = ⋃₀(↑s) :=
by simp_rw [finset.sup_eq_bUnion, set.sUnion_eq_bUnion, finset.mem_coe, id]
```

Reply:
This one is good too:
```lean
lemma finset.sup_id_eq_sUnion {α β : Type*} {s : finset (set α)} :
  s.sup id = ⋃₀(↑s) :=
by simp_rw [finset.sup_eq_bUnion, set.sUnion_eq_bUnion, finset.mem_coe, id]
```

Reply:
I can't but agree. 
A lemma named docs#finset.sup_eq_bUnion actually exists, and I inserted the second one just after it.